File: 0105. Construct Binary Tree from Preorder and Inorder Traversal/Solution0105.py
class Solution0105:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        def helper(left, right):
            nonlocal preorder_index
            if left > right: return None
            root_value = preorder[preorder_index]
            preorder_index += 1
            root = TreeNode(root_value)
            root.left = helper(left, inMap[root_value]-1)
            root.right = helper(inMap[root_value]+1, right)
            return root
        
        preorder_index = 0
        inMap = {}
        for index, value in enumerate(inorder):
            inMap[value] = index
        return helper(0, len(inorder)-1)

File: 0105. Construct Binary Tree from Preorder and Inorder Traversal/Solution0105.cpp
class Solution0105 {
public:
    int preorderIdx;
    unordered_map<int, int> inMap;
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
       for (int i = 0; i < inorder.size(); i++) {
           inMap[inorder[i]] = i;
       }
       preorderIdx = 0;
       return helper(preorder, 0, preorder.size() - 1);        
    }
    
    TreeNode* helper(vector<int>& preorder, int left, int right) {
        if (left>right) return nullptr;
        int rootVal = preorder[preorderIdx++];
        TreeNode* root = new TreeNode(rootVal);
        root->left = helper(preorder, left, inMap[rootVal]-1);
        root->right = helper(preorder, inMap[rootVal]+1, right);
        return root;
    }
};
